index.js

```diff
@@ -2900,6 +2947,14 @@
 
   return S;
 
+  };
+
+  //  Export two versions of the Sanctuary module: one with type checking;
+  //  one without.
+  var S       = createSanctuary(true);
+  S.unchecked = createSanctuary(false);
+  return S;
+
```

Comment:
Thank you for your thoughtful response, Aldwin.

> - Consumption will become weird and fiddly when doing "unchecked in production":
>   * `var S = isProduction ? require('sanctuary').unchecked : require('sanctuary')`
>   * `import sanctuary from 'sanctuary'; const S = isProduction ? sanctuary.unchecked : sanctuary`

I agree that this is fiddly. The fiddliness can, though, be contained to a single file:

```javascript
// sanctuary.js

'use strict'

const S = require('sanctuary');

const env = require('./env');

module.exports = env.isProduction ? S.unchecked : S;
```

```javascript
// app.js

'use strict';

const S = require('./sanctuary');

// ...
```

> - Somewhere, knowing that there's an unnecessary sanctuary living in memory alongside the one I use doesn't sit well with me.

I agree that this is not ideal, though I give it a low weighting relative to the other considerations.

> - It's not "future proofed" against additional options.

I very much hope we won't ever need more options. :)

> ```javascript
> var S = require('sanctuary').create(isProduction);
> ```

The following comments relate to the approach quoted above.

> - It's more recognizable: You're just instantiating the Sanctuary class, like with many other Node API's

I find this a bit confusing. Am I creating an instance of some "class"? Does this "class" have a prototype? Is the idea that I can then mutate the instance (to add my own functions, say), or is the object shared in some way?

> - It's very easy to do "unchecked in production".

I can't argue with you on this one. :)

> - It allows adding more options in the future

As mentioned above, I very much hope this will never be necessary.

Were we to add options in future, we may at that point decide to switch to accepting a single options record rather than positional arguments. This would be particularly important if there were multiple Boolean arguments, to avoid the [Boolean trap][1]. I suggest we don't let extensibility influence this decision, since what will be best at some point in the future may be incompatible with what is best now.

> - You only need one instance in memory

This is a win, but a small one in my eyes. Sanctuary is inefficient in many respects. ;)

> - It's horribly backwards incompatible

I'm happy to break backwards compatibility in exchange for a significant improvement, so this point isn't very significant in and of itself. Only when we consider two options equally good should we favour the status quo.

I'd like to add two points for consideration:

  - Were one to `require('sanctuary').create(isProduction)` in N files, one would create N instances in memory. If N > 2, this is more wasteful than creating two modules up front. If, on the other hand, all N expressions result in references to a single module, I argue that the resemblance to instantiation hinders rather than aids understanding of the function's behaviour.

  - More significant than the question of backwards compatibility is the *experience* of importing Sanctuary. Currently a user can launch a REPL, `require('sanctuary')`, and begin interacting with the library. This is how we expect most packages to behave. Doing something else would be a minor inconvenience for experienced users; a source of confusion for new users. Expecting a new user to read documentation in order to have a satisfying first experience with the library is asking a lot.

Do you find this reasoning compelling, @Avaq? What are your thoughts, @svozza? @benperez?


[1]: http://ariya.ofilabs.com/2011/08/hall-of-api-shame-boolean-trap.html